C/CasADi/build_tarballs.jl

```diff
@@ -0,0 +1,79 @@
+using BinaryBuilder, Pkg
+
+name = "CasADi"
+
+version = v"3.5.5"
+
+sources = [
+    GitSource(
+        "https://github.com/casadi/casadi.git",
+        "fadc86444f3c7ab824dc3f2d91d4c0cfe7f9dad5",
+    ),
+    DirectorySource("./bundled"),
+]
+
+script = raw"""
+cd $WORKSPACE/srcdir/casadi
+install_license LICENSE.txt
+
+mkdir -p build
+cd build
+
+export CXXFLAGS="-fPIC -std=c++11"
+export CFLAGS="${CFLAGS} -fPIC"
+
+cmake -DCMAKE_INSTALL_PREFIX=${prefix} \
+    -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
+    -DCMAKE_BUILD_TYPE=Release \
+    -DWITH_IPOPT=ON \
+    -DWITH_EXAMPLES=OFF \
+    -DWITH_DEEPBIND=OFF \
+    ..
+
+if [[ "${target}" == *-linux-* ]]; then
+        make -j ${nproc}
+else
+    cmake --build . --config Release --parallel
+fi
+make install
+
+cd $WORKSPACE/srcdir
+
+${CXX} main.cpp -o ${bindir}/amplexe -I${includedir} -L${libdir} -lcasadi -std=c++11
+"""
+
+products = [
+    ExecutableProduct("amplexe", :amplexe),
+    LibraryProduct("libcasadi", :libcasadi),
+    LibraryProduct("libcasadi_nlpsol_ipopt", :libcasadi_nlpsol_ipopt),
+]
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+platforms = supported_platforms()
+platforms = expand_cxxstring_abis(platforms)
+filter!(platforms) do p
+    # Windows export bug is:
+    #   https://gcc.gnu.org/bugzilla/show_bug.cgi?id=50044
+    # because of
+    #   https://github.com/casadi/casadi/blob/402fe583f0d3cf1fc77d1e1ac933f75d86083124/casadi/core/dm_instantiator.cpp#L373-L374
+    return !Sys.iswindows(p)
+end
```

Comment:
I'm trying to build locally for Windows.  The only problem I see is that the build system installs stuff in completely bizarre and non-sense directories, so compilation of `amplexe` fails because it can't find the header file and the shared library.  Of course figuring out where in the CMake this happens is not fun.

Notice for kids at home: this uses a ***lot** of templating, make sure you have 15+ GB of RAM completely free before attempting to build with no more than 3 threads, unless you want to OOM your system.

Reply:
Haha. I guess you should have warned you about this. I could only really debug this by pushing commits to Yggdrasil, because I hit memory problems locally and it would either crash or start paging and take forever :(

Reply:
Ok, for the non-sense installation the patch is
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 19e4d02c9..e7a1b43bf 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -314,7 +314,7 @@ option(WITH_COPYSIGN_UNDEF "On Windows, in Python.h #defines copysign to be _cop
 ############################################################
 # Flat hierarchy if self-contained and on Windows
 option(WITH_SELFCONTAINED "Make the install directory self-contained" OFF)
-if(WITH_SELFCONTAINED OR (WIN32 AND NOT CYGWIN))
+if(WITH_SELFCONTAINED)
   set(DEF_LIB_PREFIX casadi)
   set(DEF_BIN_PREFIX casadi)
   set(DEF_CMAKE_PREFIX casadi/cmake)
```
but after that, I get the same error as you
```console
sandbox:${WORKSPACE}/srcdir # ${CXX} main.cpp -o ${bindir}/amplexe -I${includedir} -L${libdir} -lcasadi -std=c++11
/tmp/ccCjdnJp.o:main.cpp:(.text$_ZNK6casadi6MatrixIdEcvSt6vectorIT_SaIS3_EEIdEEv[_ZNK6casadi6MatrixIdEcvSt6vectorIT_SaIS3_EEIdEEv]+0x61): undefined reference to `casadi::Matrix<double>::nonzeros() const'
/tmp/ccCjdnJp.o:main.cpp:(.text$_ZNK6casadi13GenericMatrixINS_6MatrixIdEEE8sparsityEv[_ZNK6casadi13GenericMatrixINS_6MatrixIdEEE8sparsityEv]+0x19): undefined reference to `casadi::Matrix<double>::sparsity() const'
collect2: error: ld returned 1 exit status
```
My understanding is that mangling has gone very badly wrong.

Reply:
> The only problem I see is that the build system installs stuff in completely bizarre and non-sense directories

Also yet. I had a bunch of hacks to move things into the right place in order to see the linking bug:
https://github.com/JuliaPackaging/Yggdrasil/pull/5048/commits/5c1d6c64f0abe22d45167d6ddc57497b1a35e0fa